components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -71,6 +74,156 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
         }
     }
 
+    #[inline]
+    pub fn on_refresh_region_buckets(
+        &mut self,
+        region_epoch: RegionEpoch,
+        mut buckets: Vec<raftstore::store::Bucket>,
+        bucket_ranges: Option<Vec<raftstore::store::BucketRange>>,
+    ) {
+        // bucket version layout
+        //   term       logical counter
+        // |-----------|-----------|
+        //  high bits     low bits
+        // term: given 10s election timeout, the 32 bit means 1362 year running time
+        let gen_bucket_version = |term, current_version| {
+            let current_version_term = current_version >> 32;
+            let bucket_version: u64 = if current_version_term == term {
+                current_version + 1
+            } else {
+                if term > u32::MAX.into() {
+                    error!(
+                        self.fsm.peer().logger,
+                        "unexpected term {} more than u32::MAX. Bucket
+                    version will be backward.",
+                        term
+                    );
+                }
+                term << 32
+            };
+            bucket_version
+        };
+
+        let region = self.fsm.peer().region();
+        if util::is_epoch_stale(&region_epoch, region.get_region_epoch()) {
+            error!(
+                 self.fsm.peer().logger,
+                "receive a stale refresh region bucket message";
+                "epoch" => ?region_epoch,
+                "current_epoch" => ?region.get_region_epoch(),
+            );
+
+            return;
+        }
+
+        let mut current_version = self
+            .fsm
+            .peer()
+            .region_buckets()
+            .as_ref()
+            .map(|b| b.meta.version)
+            .unwrap_or_default();
+        if current_version == 0 {
+            current_version = self
+                .fsm
+                .peer()
+                .last_region_buckets()
+                .as_ref()
+                .map(|b| b.meta.version)
+                .unwrap_or_default();
+        }
+        let mut region_buckets: BucketStat;
+        if let Some(bucket_ranges) = bucket_ranges {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut i = 0;
+            region_buckets = self.fsm.peer().region_buckets().clone().unwrap();
```

Comment:
How do you make sure this isn't None?

Reply:
It depends on the `bucket_ranges`, the `bucket_ranges` should be generated by the `region_buckets`. So if `bucket_ranges` is none, it means that the `region_buckets` is None.

Reply:
That's too obscure. I have to trace all the way `on_refresh_region_buckets -> PeerMsg::RefreshRegionBuckets -> refresh_region_buckets -> approximate_check_bucket/scan_split_keys -> ...` to know where the message is coming from. And as you said it before, `region_buckets` can be reset. The ordering between this function and the reset is hard to reason.

Reply:
yes, I will conside the condition.